src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1743,6 +1758,35 @@ string dumpDagTest(BoundDagTest d)
 #endif
         }
 
+        /// <summary>
+        /// This is a readonly wrapper around an array builder.  It ensures we can benefit from the pooling an array builder provides, without having to incur 
+        /// intermediary allocations for <see cref="ImmutableArray"/>s.
+        /// </summary>
+        private readonly struct FrozenArrayBuilder<T>
+        {
+            private readonly ArrayBuilder<T> _arrayBuilder;
+
+            public FrozenArrayBuilder(ArrayBuilder<T> arrayBuilder)
+            {
+                _arrayBuilder = arrayBuilder;
+            }
+
+            public bool IsDefault => _arrayBuilder is null;
+
+            public void Free()
+            {
+                _arrayBuilder.Free();
```

Comment:
`_arrayBuilder` could be `null`

Reply:
Not in legal code.  This follows the pooled object-pattern where the nullable annotations reflect the state when things are used validly.  In the pool you should never see/use them, so annotating for that case just means warnings for thigns that shoudl not happen (and which you'd want to blow up).